pkg/controller/clone/planner.go

```diff
@@ -0,0 +1,534 @@
+		args.Log.V(3).Info("Source PVC does not exist, cannot compute strategy")
+		return nil, nil
+	}
+
+	if err = p.validateSourcePVC(args, sourceClaim); err != nil {
+		p.Recorder.Event(args.TargetClaim, corev1.EventTypeWarning, CloneValidationFailed, MessageCloneValidationFailed)
+		args.Log.V(1).Info("Validation failed", "target", args.TargetClaim, "source", sourceClaim)
+		return nil, err
+	}
+
+	strategy := cdiv1.CloneStrategySnapshot
+	cs, err := GetGlobalCloneStrategyOverride(ctx, p.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	if cs != nil {
+		strategy = *cs
+	} else if args.TargetClaim.Spec.StorageClassName != nil {
+		sp := &cdiv1.StorageProfile{}
+		exists, err := getResource(ctx, p.Client, "", *args.TargetClaim.Spec.StorageClassName, sp)
+		if err != nil {
+			return nil, err
+		}
+
+		if !exists {
+			args.Log.V(3).Info("missing storageprofile for", "name", *args.TargetClaim.Spec.StorageClassName)
+		}
+
+		if exists && sp.Status.CloneStrategy != nil {
+			strategy = *sp.Status.CloneStrategy
+		}
+	}
+
+	if strategy == cdiv1.CloneStrategySnapshot {
+		n, err := GetCompatibleVolumeSnapshotClass(ctx, p.Client, sourceClaim, args.TargetClaim)
+		if err != nil {
+			return nil, err
+		}
+
+		if n == nil {
+			strategy = cdiv1.CloneStrategyHostAssisted
+		}
+	}
+
+	if strategy == cdiv1.CloneStrategySnapshot ||
+		strategy == cdiv1.CloneStrategyCsiClone {
+		ok, err := p.validateAdvancedClonePVC(ctx, args, sourceClaim)
+		if err != nil {
+			return nil, err
+		}
+
+		if !ok {
+			strategy = cdiv1.CloneStrategyHostAssisted
+		}
+	}
+
+	return &strategy, nil
+}
+
+func (p *Planner) validateTargetStorageClassAssignment(ctx context.Context, args *ChooseStrategyArgs) (bool, error) {
+	if args.TargetClaim.Spec.StorageClassName == nil {
+		args.Log.V(3).Info("Target PVC has nil storage class, cannot compute strategy")
+		return false, nil
+	}
+
+	if *args.TargetClaim.Spec.StorageClassName == "" {
+		args.Log.V(3).Info("Target PVC has \"\" storage class, cannot compute strategy")
+		return false, fmt.Errorf("claim has emptystring storageclass, will not work")
+	}
+
+	sc, err := GetStorageClassForClaim(ctx, p.Client, args.TargetClaim)
+	if err != nil {
+		return false, err
+	}
+
+	if sc == nil {
+		args.Log.V(3).Info("Target PVC has no storage class, cannot compute strategy")
+		return false, fmt.Errorf("target storage class not found")
+	}
+
+	return true, nil
+}
+
+func (p *Planner) validateSourcePVC(args *ChooseStrategyArgs, sourceClaim *corev1.PersistentVolumeClaim) error {
+	if err := cc.ValidateRequestedCloneSize(sourceClaim.Spec.Resources, args.TargetClaim.Spec.Resources); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Planner) validateAdvancedClonePVC(ctx context.Context, args *ChooseStrategyArgs, sourceClaim *corev1.PersistentVolumeClaim) (bool, error) {
+	if !SameVolumeMode(sourceClaim, args.TargetClaim) {
+		args.Log.V(3).Info("volume modes not compatible for advanced clone")
+		return false, nil
```

Comment:
Does it make sense to emit an event here indicating why we switched to host assisted clone? In most cases we are not logging at V(3) and this piece of information gets lost. I would settle for a V(1) log honestly.